### DISCUSSION TITLE
Using Matter.js as physics engine

### DISCUSSION THREAD

**Discussion by justinfernald** [excaliburjs/Excalibur#2659] [Q&A] [answered]
I am trying to figure out how I would use the Matter.js physics engine instead of the built in realistic physics engine. The reason being that a big part of a game I would like to make is highly dependent on [constraints](https://brm.io/matter-js/docs/classes/Constraint.html). If someone could point me in the direction of how this might be possible, if it is possible that would be great.

**Comment by eonarheim** [comment, marked as answer]
@justinfernald This should be possible to do!

I haven't attempted this yet (I'll have some time later this week), but the approach I'd try is using [Excalibur's ECS](https://excaliburjs.com/docs/entity-component-system/):

1. Build a custom ECS `System` to host the matter.js code [Engine/World and run the simulation](https://github.com/liabru/matter-js/wiki/Getting-started). It looks like matter.js supports manually [running thesimulation](https://github.com/liabru/matter-js/wiki/Running). 
  a. Here is a simple-ish system in the docs https://excaliburjs.com/docs/systems/#implementing-your-own-components--systems

3. Create new a new ECS component type to wrap the matter.js Body/Constraints. When this new component is added to an Excalibur actor that matches a system's types it [notifies the system](https://github.com/excaliburjs/Excalibur/blob/main/src/engine/Collision/CollisionSystem.ts#L39). This notify mechanism in the new system is where you could wire into matter.js

4. Sync the matter.js body positions back to the Excalibur actor transforms after you step the matter.js simulation 

I'll let you know what I come up with this week! I'm excited to try this!

**Reply by eonarheim** [reply]
Oh you might also want to disable excalibur's physics as well https://excaliburjs.com/docs/physics/#other-physics-settings

`ex.Physics.enabled = false`

**Reply by justinfernald** [reply]
Thank you for the response @eonarheim!

**Reply by eonarheim** [reply]
@justinfernald I just posted an example! https://github.com/excaliburjs/sample-matterjs
